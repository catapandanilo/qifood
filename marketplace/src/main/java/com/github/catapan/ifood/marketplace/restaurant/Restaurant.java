package com.github.catapan.ifood.marketplace.restaurant;

import com.github.catapan.ifood.marketplace.localization.Localization;

public class Restaurant {

  public Long id;

  public String name;

  public Localization localization;
}
