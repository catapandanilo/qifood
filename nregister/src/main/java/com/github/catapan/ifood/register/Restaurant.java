package com.github.catapan.ifood.register;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;
import java.util.Date;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.Table;
import org.hibernate.annotations.CreationTimestamp;

@Entity
@Table(name = "restaurant")
public class Restaurant extends PanacheEntityBase {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    public String owner;

    public String cnpj;

    public String name;

    @ManyToOne
    public Localization localization;

    @CreationTimestamp
    public Date dateCreation;

    @CreationTimestamp
    public Date dataUpdate;
}
