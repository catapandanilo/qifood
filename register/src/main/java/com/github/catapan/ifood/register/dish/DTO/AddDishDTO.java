package com.github.catapan.ifood.register.dish.DTO;

import java.math.BigDecimal;

public class AddDishDTO {

  public String name;

  public String description;

  public BigDecimal price;
}
