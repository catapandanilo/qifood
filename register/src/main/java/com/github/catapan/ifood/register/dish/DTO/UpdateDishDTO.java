package com.github.catapan.ifood.register.dish.DTO;

import java.math.BigDecimal;

public class UpdateDishDTO {

  public BigDecimal price;
}
