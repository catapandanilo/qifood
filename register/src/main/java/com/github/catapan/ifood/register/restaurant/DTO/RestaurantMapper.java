package com.github.catapan.ifood.register.restaurant.DTO;

import com.github.catapan.ifood.register.restaurant.Restaurant;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;

@Mapper(componentModel = "cdi")
public interface RestaurantMapper {

  @Mapping(target = "name", source = "fantasyName")
  @Mapping(target = "id", ignore = true)
  @Mapping(target = "dateCreation", ignore = true)
  @Mapping(target = "dateUpdate", ignore = true)
  @Mapping(target = "localization.id", ignore = true)
  public Restaurant toRestaurant(AddRestaurantDTO dto);

  @Mapping(target = "name", source = "fantasyName")
  public void toRestaurant(UpdateRestaurantDTO updateRestaurantDTO,
    @MappingTarget Restaurant restaurant);

  @Mapping(target = "fantasyName", source = "name")
  @Mapping(target = "dateCreation", dateFormat = "dd/MM/yyyy HH:mm:ss")
  public RestaurantDTO toRestaurantDTO(Restaurant restaurant);
}
