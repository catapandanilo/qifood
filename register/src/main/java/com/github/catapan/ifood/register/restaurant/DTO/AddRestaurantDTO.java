package com.github.catapan.ifood.register.restaurant.DTO;

import com.github.catapan.ifood.register.localization.DTO.LocalizationDTO;
import com.github.catapan.ifood.register.restaurant.Restaurant;
import com.github.catapan.ifood.register.restaurant.infra.DTO;
import javax.validation.ConstraintValidatorContext;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

public class AddRestaurantDTO implements DTO {

  @Pattern(regexp = "[0-9]{2}\\.[0-9]{3}\\.[0-9]{3}\\/[0-9]{4}\\-[0-9]{2}")
  @NotNull
  public String cnpj;

  @Size(min = 3, max = 30)
  public String fantasyName;

  public LocalizationDTO localization;

  @Override
  public boolean isValid(ConstraintValidatorContext constraintValidatorContext) {
    constraintValidatorContext.disableDefaultConstraintViolation();
    if (Restaurant.find("cnpj", cnpj).count() > 0) {
      constraintValidatorContext.buildConstraintViolationWithTemplate("CNPJ already registered!")
        .addPropertyNode("cnpj")
        .addConstraintViolation();
      return false;
    }
    return true;
  }

}
