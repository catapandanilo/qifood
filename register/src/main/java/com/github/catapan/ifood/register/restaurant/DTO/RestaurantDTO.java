package com.github.catapan.ifood.register.restaurant.DTO;

import com.github.catapan.ifood.register.localization.DTO.LocalizationDTO;

public class RestaurantDTO {

  public Long id;

  public String owner;

  public String cnpj;

  public String fantasyName;

  public LocalizationDTO localization;

  public String dateCreation;

}
