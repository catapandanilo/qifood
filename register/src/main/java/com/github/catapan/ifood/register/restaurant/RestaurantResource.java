package com.github.catapan.ifood.register.restaurant;

import com.github.catapan.ifood.register.restaurant.DTO.AddRestaurantDTO;
import com.github.catapan.ifood.register.restaurant.DTO.RestaurantDTO;
import com.github.catapan.ifood.register.restaurant.DTO.RestaurantMapper;
import com.github.catapan.ifood.register.restaurant.DTO.UpdateRestaurantDTO;
import com.github.catapan.ifood.register.restaurant.infra.ConstraintViolationResponse;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.validation.Valid;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;
import org.eclipse.microprofile.metrics.annotation.Counted;
import org.eclipse.microprofile.metrics.annotation.SimplyTimed;
import org.eclipse.microprofile.metrics.annotation.Timed;
import org.eclipse.microprofile.openapi.annotations.enums.SecuritySchemeType;
import org.eclipse.microprofile.openapi.annotations.media.Content;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlow;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlows;
import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirement;
import org.eclipse.microprofile.openapi.annotations.security.SecurityScheme;
import org.eclipse.microprofile.openapi.annotations.tags.Tag;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;

@Path("/restaurants")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
//@RolesAllowed("owner")
@Tag(name = "Restaurant")
@SecurityScheme(securitySchemeName = "ifood-oauth", type = SecuritySchemeType.OAUTH2, flows = @OAuthFlows(password = @OAuthFlow(tokenUrl = "http://localhost:8180/auth/realms/ifood/protocol/openid-connect/token")))
@SecurityRequirement(name = "ifood-oauth", scopes = {})
public class RestaurantResource {

  @Inject
  RestaurantMapper restaurantMapper;

  @Inject
  @Channel("restaurants")
  Emitter<Restaurant> emitter;

  @GET
  @Counted(
	name = "getAll/Counted",
	description = "Quantity searches restaurants"
)
  @SimplyTimed(
	name = "getAll/SimplyTimed",
	description = "Simple time to searches"
)
  @Timed(
	name = "getAll/Timed",
	description = "Complete time to searches"
)
  public List<RestaurantDTO> getAll() {
    Stream<Restaurant> restaurants = Restaurant.streamAll();
    return restaurants.map(restaurant -> restaurantMapper.toRestaurantDTO(restaurant))
      .collect(Collectors.toList());
  }

  @POST
  @Transactional
  @APIResponse(responseCode = "201", description = "When the restaurant is successfully registered")
  @APIResponse(responseCode = "400", content = @Content(schema = @Schema(allOf = ConstraintViolationResponse.class)))
  public Response add(@Valid AddRestaurantDTO addRestaurantDTO) {
    Restaurant restaurant = restaurantMapper.toRestaurant(addRestaurantDTO);
    restaurant.persist();

//    Jsonb create = JsonbBuilder.create();
//    String json = create.toJson(restaurant);
    emitter.send(restaurant);

    return Response.status(Status.CREATED).build();
  }

  @PUT
  @Path("{id}")
  @Transactional
  public void update(@PathParam("id") Long id, UpdateRestaurantDTO updateRestaurantDTO) {
    Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(id);
    if (restaurantOptional.isEmpty()) {
      throw new NotFoundException();
    }

    Restaurant restaurantUpdated = restaurantOptional.get();
    restaurantMapper.toRestaurant(updateRestaurantDTO, restaurantUpdated);
    restaurantUpdated.persist();
  }

  @DELETE
  @Path("{id}")
  @Transactional
  public void delete(@PathParam("id") Long id) {
    Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(id);
    restaurantOptional.ifPresentOrElse(Restaurant::delete, () -> {
      throw new NotFoundException();
    });
  }
}