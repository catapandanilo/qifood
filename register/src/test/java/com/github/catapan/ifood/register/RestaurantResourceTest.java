package com.github.catapan.ifood.register;

import static io.restassured.RestAssured.given;

import com.github.catapan.ifood.register.restaurant.DTO.UpdateRestaurantDTO;
import com.github.catapan.ifood.register.restaurant.Restaurant;
import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.specification.RequestSpecification;
import org.junit.Assert;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import javax.ws.rs.core.Response.Status;
import org.approvaltests.Approvals;

@DBRider
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE)
@QuarkusTest
@QuarkusTestResource(RegisterTestLifecycleManager.class)
public class RestaurantResourceTest {

  @Test
  @DataSet("restaurants-scenery-1.yml")
  public void testGetRestaurants() {
    String result = given()
      .when().get("/restaurants")
      .then()
      .statusCode(Status.OK.getStatusCode())
      .extract().asString();
    Approvals.verifyJson(result);
  }

  private RequestSpecification given() {
    return RestAssured.given().contentType(ContentType.JSON);
  }

  @Test
  @DataSet("restaurants-scenery-1.yml")
  public void testUpdateRestaurant() {
    UpdateRestaurantDTO updateRestaurantDTO = new UpdateRestaurantDTO();
    updateRestaurantDTO.fantasyName = "newName";
    Long parameterValue = 123L;
    given()
      .with().pathParam("id", parameterValue)
      .body(updateRestaurantDTO)
      .when().put("/restaurants/{id}")
      .then()
      .statusCode(Status.NO_CONTENT.getStatusCode())
      .extract().asString();

    Restaurant findById = Restaurant.findById(parameterValue);

    Assertions.assertEquals(updateRestaurantDTO.fantasyName, findById.name);
  }
}
