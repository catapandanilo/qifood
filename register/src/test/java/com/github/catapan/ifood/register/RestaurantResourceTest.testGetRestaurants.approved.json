[
  {
    "fantasyName": "New Restaurant",
    "id": 123,
    "owner": "ID do keyclock"
  },
  {
    "fantasyName": "New Restaurant 2",
    "id": 234,
    "owner": "ID do keyclock 2"
  }
]
